File: java/02-intermediate/02-inheritance/InheritanceDemo.java
/**
 * LESSON: Inheritance
 * A class can inherit fields and methods from another class using 'extends'.
 * Promotes code reuse and establishes an IS-A relationship.
 */
public class InheritanceDemo {
    public static void main(String[] args) {
        Dog dog = new Dog("Buddy", 3, "Golden Retriever");

        System.out.println("--- Dog Info ---");
        System.out.println("Name: " + dog.getName());
        System.out.println("Age: " + dog.getAge());
        System.out.println("Breed: " + dog.getBreed());

        // Inherited method
        dog.eat();

        // Overridden method
        dog.makeSound();

        // Dog's own method
        dog.fetch();

        // Parent reference can hold child object (upcasting)
        System.out.println("\n--- Upcasting ---");
        Animal animal = new Dog("Rex", 5, "Labrador");
        animal.eat();       // inherited method
        animal.makeSound(); // calls DOG's version (runtime polymorphism)
        // animal.fetch();  // ERROR: Animal reference can't see Dog-specific methods
    }
}

// Parent (Super) class
class Animal {
    private String name;
    private int age;

    Animal(String name, int age) {
        this.name = name;
        this.age = age;
        System.out.println("[Animal constructor called]");
    }

    void eat() {
        System.out.println(name + " is eating.");
    }

    void makeSound() {
        System.out.println(name + " makes a sound.");
    }

    String getName() { return name; }
    int getAge() { return age; }
}

// Child (Sub) class
class Dog extends Animal {
    private String breed;

    Dog(String name, int age, String breed) {
        super(name, age); // call parent constructor FIRST
        this.breed = breed;
        System.out.println("[Dog constructor called]");
    }

    // Method overriding - providing a specific implementation
    @Override
    void makeSound() {
        System.out.println(getName() + " says: Woof! Woof!");
    }

    void fetch() {
        System.out.println(getName() + " is fetching the ball!");
    }

    String getBreed() { return breed; }
}

File: cpp-practice/02-intermediate/02-inheritance/inheritance_demo.cpp
/**
 * LESSON: Inheritance
 * Derive new classes from existing ones, reusing and extending functionality.
 *
 * Access specifiers for inheritance:
 *   public:    public->public, protected->protected
 *   protected: public->protected, protected->protected
 *   private:   public->private, protected->private
 */
#include <iostream>
#include <string>

// Base (Parent) class
class Animal {
protected:
    std::string name;
    int age;

public:
    Animal(const std::string& name, int age) : name(name), age(age) {
        std::cout << "[Animal constructor: " << name << "]\n";
    }

    virtual ~Animal() {
        std::cout << "[Animal destructor: " << name << "]\n";
    }

    void eat() const {
        std::cout << name << " is eating.\n";
    }

    virtual void makeSound() const {
        std::cout << name << " makes a generic sound.\n";
    }

    std::string getName() const { return name; }
    int getAge() const { return age; }
};

// Derived (Child) class
class Dog : public Animal {
private:
    std::string breed;

public:
    Dog(const std::string& name, int age, const std::string& breed)
        : Animal(name, age), breed(breed) {
        std::cout << "[Dog constructor: " << name << "]\n";
    }

    ~Dog() override {
        std::cout << "[Dog destructor: " << name << "]\n";
    }

    // Override parent method
    void makeSound() const override {
        std::cout << name << " says: Woof! Woof!\n";
    }

    void fetch() const {
        std::cout << name << " is fetching the ball!\n";
    }

    std::string getBreed() const { return breed; }
};

// Multilevel: Animal -> Dog -> Puppy
class Puppy : public Dog {
public:
    Puppy(const std::string& name, const std::string& breed)
        : Dog(name, 0, breed) {}

    void makeSound() const override {
        std::cout << name << " says: Yip! Yip!\n";
    }

    void play() const {
        std::cout << name << " is playing!\n";
    }
};

int main() {
    std::cout << "--- Creating Dog ---\n";
    Dog dog("Buddy", 3, "Golden Retriever");
    dog.eat();       // inherited
    dog.makeSound(); // overridden
    dog.fetch();     // own method

    std::cout << "\n--- Upcasting (polymorphism) ---\n";
    Animal* animalPtr = &dog; // parent pointer to child object
    animalPtr->eat();
    animalPtr->makeSound(); // calls Dog's version (virtual dispatch)
    // animalPtr->fetch();  // ERROR: Animal doesn't know about fetch()

    std::cout << "\n--- Multilevel Inheritance ---\n";
    Puppy puppy("Max", "Labrador");
    puppy.eat();       // from Animal
    puppy.makeSound(); // from Puppy
    puppy.fetch();     // from Dog
    puppy.play();      // own

    std::cout << "\n--- Destructors ---\n";
    return 0;
}
